File: Sources/Model/UsageBand.swift
import SwiftUI

/// The colour a ring or bar takes at a given level of use.
///
/// The thresholds come from the mockup, which shows 21% green, 52% yellow and
/// 73% orange. (The prose table in the design spec says 50–79 is yellow, which
/// would make 73% yellow and contradict the frame it claims to describe — the
/// frame wins.)
enum UsageBand: Equatable {
    case ample       // under half
    case watch       // getting close
    case critical    // nearly out
    case exhausted   // limit hit, waiting for the reset

    static func band(for usedFraction: Double) -> UsageBand {
        switch usedFraction {
        case ..<0.50: return .ample
        case ..<0.70: return .watch
        case ..<1.00: return .critical
        default:      return .exhausted
        }
    }

    /// `accent` only ever stands in for the ample state's colour — the
    /// warning bands stay fixed regardless of the chosen accent, since their
    /// whole job is to interrupt whatever else is on screen and a
    /// customisable warning colour could be tuned into invisibility.
    func color(accent: Color = Palette.ample) -> Color {
        switch self {
        case .ample:                 return accent
        case .watch:                 return Palette.watch
        case .critical, .exhausted:  return Palette.critical
        }
    }
}

File: windows/QuotaArc/Model/UsageBand.cs
using System.Windows.Media;
using QuotaArc.Design;
using QuotaArc.Settings;

namespace QuotaArc.Model;

internal enum UsageBand
{
    Ample,
    Watch,
    Critical,
    Exhausted
}

internal static class UsageBands
{
    public static UsageBand Band(double usedFraction) => usedFraction switch
    {
        < 0.50 => UsageBand.Ample,
        < 0.70 => UsageBand.Watch,
        < 1.00 => UsageBand.Critical,
        _ => UsageBand.Exhausted
    };

    public static Color Color(this UsageBand band, AccentColorChoice accent = AccentColorChoice.System) => band switch
    {
        UsageBand.Ample => accent.Color(),
        UsageBand.Watch => Palette.Watch,
        _ => Palette.Critical
    };

    public static Brush Brush(this UsageBand band, AccentColorChoice accent = AccentColorChoice.System)
    {
        if (band == UsageBand.Ample) return accent.Brush();
        return band == UsageBand.Watch ? Palette.WatchBrush : Palette.CriticalBrush;
    }
}
